### PR TITLE
Remove public BigInteger references and use explicit operators

### PR DESCRIPTION
**Pull request by quantumagi**

### CODE REVIEW

File: Stratis.SmartContracts/UInt256.cs [hunk 1]
```diff
@@ -157,39 +157,44 @@ public UInt256(byte[] vch) : this(vch, true)
...
-        public static implicit operator ulong(UInt256 value)
+        public static explicit operator ulong(UInt256 value)
         {
             return (ulong)value.value.GetValue();
         }
 
-        public static implicit operator BigInteger(UInt256 value)
+        public static explicit operator UInt128(UInt256 value)
         {
-            return value.value.GetValue();
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UInt256.cs:192]: Can this stay as `new UInt128(value.value.GetValue())` and make private constructor as internal ?

File: Stratis.SmartContracts/UInt256.cs [hunk 2]
```diff
@@ -157,39 +157,44 @@ public UInt256(byte[] vch) : this(vch, true)
...
         }
 
-        public static implicit operator UInt256(BigInteger value)
-        {
-            return new UInt256(value);
-        }
-
         public static implicit operator UInt256(UInt128 value)
         {
-            return new UInt256(value);
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UInt256.cs:167]: `new UInt256(value.ToBytes())` ?

File: Stratis.SmartContracts/UInt256.cs [hunk 3]
```diff
@@ -157,39 +157,44 @@ public UInt256(byte[] vch) : this(vch, true)
             return new UInt256(value);
         }
 
-        public static implicit operator UInt256(BigInteger value)
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UInt256.cs:160]: This can stay as internal if it will simplify some code

File: Stratis.SmartContracts/UInt256.cs [hunk 4]
```diff
@@ -23,7 +23,7 @@ public static UInt256 Parse(string str)
             return new UInt256(str);
         }
 
-        public UInt256(BigInteger value)
+        private UInt256(BigInteger value)
```
Comment (YakupIpek) [inline Stratis.SmartContracts/UInt256.cs]: Internal is better probably based on code changes

File: Stratis.SmartContracts/UInt256.cs (hunk 3 above)
Comment (quantumagi) [reply Stratis.SmartContracts/UInt256.cs:160]: I thought the same but unfortunately "internal" is not accepted here.

File: Stratis.SmartContracts/UInt256.cs (hunk 2 above)
Comment (quantumagi) [reply Stratis.SmartContracts/UInt256.cs:167]: Byte array length must be correct.

File: Stratis.SmartContracts/UInt256.cs (hunk 1 above)
Comment (quantumagi) [reply Stratis.SmartContracts/UInt256.cs:192]: Done.

File: Stratis.SmartContracts/UInt256.cs (hunk 4 above)
Comment (quantumagi) [reply Stratis.SmartContracts/UInt256.cs]: Yes.

File: Stratis.SmartContracts/UInt256.cs (hunk 3 above)
Comment (YakupIpek) [reply Stratis.SmartContracts/UInt256.cs:160]: you can convert it to a internal public method actually.

`internal static ToUInt256(BigInteger value)`

File: Stratis.SmartContracts/UInt256.cs (hunk 2 above)
Comment (YakupIpek) [reply Stratis.SmartContracts/UInt256.cs:167]: suggested `ToUInt256` can be used here

**Review by YakupIpek — CHANGES_REQUESTED** [review]
Can these test added for UInt128 as well ?

**Review by fassadlr — APPROVED** [review]
@YakupIpek Please also review 👍

**Review by YakupIpek — COMMENTED** [review]
@fassadlr It looks good to me.

**Review by YakupIpek — APPROVED** [review]

**Comment by YakupIpek** [comment]
Just one request can we rename the types as lowercase for uint256 and uint128 like other numeric types ?  This way developers recognize these types are special and supported as parameter on contract calls ? It seems nbitcoin also use lowercase convention.

This request can go another pr.

Wdyt ? Skip it if you disagree.

**Comment by YakupIpek** [comment]
@fassadlr This is ready to merge and publish it on nuget

**Pull request merged**